// passkit-generator/lib/cjs/schemas/SemanticTagType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WifiNetwork = exports.Seat = exports.PersonNameComponents = exports.Location = exports.EventDateInfo = exports.CurrencyAmount = void 0;
const tslib_1 = require("tslib");
const joi_1 = tslib_1.__importDefault(require("joi"));
const regexps_js_1 = require("./regexps.js");
exports.CurrencyAmount = joi_1.default.object().keys({
    currencyCode: joi_1.default.string(),
    amount: joi_1.default.string(),
});
exports.EventDateInfo = joi_1.default.object().keys({
    date: joi_1.default.string().isoDate().required(),
    ignoreTimeComponents: joi_1.default.boolean(),
    timeZone: joi_1.default.string(),
});
exports.Location = joi_1.default.object().keys({
    latitude: joi_1.default.number().required(),
    longitude: joi_1.default.number().required(),
});
exports.PersonNameComponents = joi_1.default.object().keys({
    givenName: joi_1.default.string(),
    familyName: joi_1.default.string(),
    middleName: joi_1.default.string(),
    namePrefix: joi_1.default.string(),
    nameSuffix: joi_1.default.string(),
    nickname: joi_1.default.string(),
    phoneticRepresentation: joi_1.default.string(),
});
exports.Seat = joi_1.default.object().keys({
    seatSection: joi_1.default.string(),
    seatRow: joi_1.default.string(),
    seatNumber: joi_1.default.string(),
    seatIdentifier: joi_1.default.string(),
    seatType: joi_1.default.string(),
    seatDescription: joi_1.default.string(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     */
    seatAisle: joi_1.default.string(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     */
    seatLevel: joi_1.default.string(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     */
    seatSectionColor: joi_1.default.string().regex(regexps_js_1.RGB_HEX_COLOR_REGEX),
});
exports.WifiNetwork = joi_1.default.object().keys({
    password: joi_1.default.string().required(),
    ssid: joi_1.default.string().required(),
});
